### ISSUE TITLE
VK pods fail with CrashLoopBackOff

### ISSUE THREAD

**Issue by giggio** [virtual-kubelet/virtual-kubelet#124]
### Environment summary

Provider: Azure
Version latest
K8s Master Info: AKS, 1.9.6
Install Method: az

### Issue Details

After running

```bash
az aks install-connector -g $nomeDoGrupoDeRecursos -n $nomeCluster --connector-name $nomeDoConector --os-type Both
```
The pods fail to start and stay on `CrashLoopBackOff`. They keep restarting but never come back.

The log shows:

```
$ kubectl logs aksconnectorgiggio1-linux-virtual-kubelet-for-aks-69c844fckwr8b
container_linux.go:247: starting container process caused "process_linux.go:359: container init caused \"rootfs_linux.go:54: mounting \\\"/etc/kubernetes/azure.json\\\" to rootfs \\\"/var/lib/docker/overlay2/9478f9bf673655d932c05b72e0e9a1a6fefde38c1321dc85a9aad73e4f9c937b/merged\\\" at \\\"/var/lib/docker/overlay2/9478f9bf673655d932c05b72e0e9a1a6fefde38c1321dc85a9aad73e4f9c937b/merged/etc/virtual-kubelet/acs.json\\\" caused \\\"not a directory\\\"\""
```

### Repo Steps

Create a 1.9.6 aks cluster, init helm, and try to install the connector with az.

**Comment by shahiddev** [comment]
Same error trying to install only the os-type Windows

**Comment by dewalds** [comment]
Also getting it since upgrading cluster to 1.9.6

**Comment by arjanschaaf** [comment]
Ran into this on a new cluster running 1.9.6 in westeurope as well. Deleted this cluster and recreated a new cluster with kubernetes version 1.8.10 => same problem.

```container init caused \"rootfs_linux.go:54: mounting \\\"/etc/kubernetes/azure.json\\\" to rootfs \\\"/var/lib/docker/overlay2/495e5449e0981f163f9f8003e01f8c02e5f30441506530aeeab36854d9ace03d/merged\\\" at \\\"/var/lib/docker/overlay2/495e5449e0981f163f9f8003e01f8c02e5f30441506530aeeab36854d9ace03d/merged/etc/virtual-kubelet/acs.json\\\" caused \\\"not a directory\\\"\""`

**Comment by robbiezhang** [comment]
It looks like a regression in k8s 1.9.6 that it doesn't allow to mount a single file.

**Comment by arjanschaaf** [comment]
@robbiezhang : I've just tested this to confirm that it is a regression problem on k8s 1.8.x. I've replaced my 1.8.10 cluster with a new 1.8.7 cluster and then I'm able to launch the virtual-kubelet pod successfully. So it is also a regression problem between 1.8.7 & 1.8.10

**Comment by robbiezhang** [comment]
@arjanschaaf, thanks for the info! A workaround for the VK is to mount the /etc/kubernetes folder from the host to the VK pod at /etc/virtual-kubelet/acs, and set env "ACS_CREDENTIAL_LOCATION" to /etc/virtual-kubelet/acs/azure.json
```
...
      env:
        - name: ACS_CREDENTIAL_LOCATION
          value: /etc/virtual-kubelet/acs/azure.json
...
      volumeMounts:
        - name: acs-credential
          mountPath: "/etc/virtual-kubelet/acs"
...
      volumes:
      - name: acs-credential
        hostPath:
          path: /etc/kubernetes
```

I can update the helm chart soon with this.

**Comment by robbiezhang** [comment]
Fix: #126

**Comment by arjanschaaf** [comment]
@robbiezhang : Thanks for the quick fix! Is your workaround a fix for something that should really be fixed in Kubernetes upstream because somehow in 1.8.x & 1.9.x they introduced an unwanted regression bug? Or is it rightfully so that the old approach of mounting the file instead of the folder isn't allowed anymore by Kubernetes? In case of a kubernetes bug, maybe an issue should be filed in upstream Kubernetes instead of only working around the problem in virtual-kubelet?

**Comment by robbiezhang** [comment]
@arjanschaaf : thanks for the comments! I agree with your point. I'm pretty sure it's a regression bug in the k8s upstream, and create an issue there: https://github.com/kubernetes/kubernetes/issues/62417. 
I can repo the issue by using the following simple spec, and the issue only happens when the hostPath file is mounted to the directory mounted by the secret volume. If the directory is mounted by an emptyDir volume, it works fine.

```
apiVersion: v1
kind: Secret
metadata:
  name: nginx-secret
type: Opaque
data:
  aaa: TUlTU0lORw==

---

apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - image: nginx
    name: nginx
    ports:
    - containerPort: 80
      name: http
      protocol: TCP
    volumeMounts:
    - mountPath: /var/log/lala
      name: secret-dir
    - mountPath: /var/log/lala/aaa.log
      name: host-file
  volumes:
  - name: secret-dir
    secret:
      secretName: nginx-secret
  - name: host-file
    hostPath:
      path: /var/log/waagent.log
      type: File
```

**Comment by robbiezhang** [comment]
Fixed in #126

**Comment by giggio** [comment]
I just tested it and the connector pods are up! Tks!
